--- ../../../llvm_git/llvm/lib/Support/CommandLine.cpp	2018-12-22 13:28:05.000000000 -0700
+++ ./src/Support/CommandLine.cpp	2018-12-22 14:03:44.000000000 -0700
@@ -25,7 +25,7 @@
 #include "llvm/ADT/SmallString.h"
 #include "llvm/ADT/StringExtras.h"
 #include "llvm/ADT/StringMap.h"
-#include "llvm/ADT/Triple.h"
+//#include "llvm/ADT/Triple.h"
 #include "llvm/ADT/Twine.h"
 #include "llvm/Config/config.h"
 #include "llvm/Support/ConvertUTF.h"
@@ -1105,8 +1105,7 @@
   BumpPtrAllocator A;
   StringSaver Saver(A);
   ExpandResponseFiles(Saver,
-         Triple(sys::getProcessTriple()).isOSWindows() ?
-         cl::TokenizeWindowsCommandLine : cl::TokenizeGNUCommandLine,
+         cl::TokenizeGNUCommandLine,
          newArgv);
   argv = &newArgv[0];
   argc = static_cast<int>(newArgv.size());
@@ -1620,7 +1619,7 @@
 
 bool parser<float>::parse(Option &O, StringRef ArgName, StringRef Arg,
                           float &Val) {
-  double dVal;
+  double dVal = 0.0;
   if (parseDouble(O, Arg, dVal))
     return true;
   Val = (float)dVal;
