--- config.h	2013-10-19 10:20:36.000000000 -0700
+++ config.h	2013-10-23 12:29:44.000000000 -0700
@@ -35,7 +35,7 @@
 #define GCC_INSTALL_PREFIX ""
 
 /* Define to 1 if you have the `arc4random' function. */
-#define HAVE_ARC4RANDOM 1
+//#define HAVE_ARC4RANDOM 1
 
 /* Define to 1 if you have the `argz_append' function. */
 /* #undef HAVE_ARGZ_APPEND */
@@ -250,7 +250,7 @@
 #define HAVE_LONGJMP 1
 
 /* Define to 1 if you have the <mach/mach.h> header file. */
-#define HAVE_MACH_MACH_H 1
+//#define HAVE_MACH_MACH_H 1
 
 /* Define to 1 if you have the <mach-o/dyld.h> header file. */
 #define HAVE_MACH_O_DYLD_H 1
@@ -262,10 +262,10 @@
 /* #undef HAVE_MALLOC_H */
 
 /* Define to 1 if you have the <malloc/malloc.h> header file. */
-#define HAVE_MALLOC_MALLOC_H 1
+//#define HAVE_MALLOC_MALLOC_H 1
 
 /* Define to 1 if you have the `malloc_zone_statistics' function. */
-#define HAVE_MALLOC_ZONE_STATISTICS 1
+//#define HAVE_MALLOC_ZONE_STATISTICS 1
 
 /* Define to 1 if you have the <memory.h> header file. */
 #define HAVE_MEMORY_H 1
@@ -445,7 +445,7 @@
 #define HAVE_SYS_WAIT_H 1
 
 /* Define if the setupterm() function is supported this platform. */
-#define HAVE_TERMINFO 1
+//#define HAVE_TERMINFO 1
 
 /* Define to 1 if you have the <termios.h> header file. */
 #define HAVE_TERMIOS_H 1
@@ -466,7 +466,7 @@
 /* #undef HAVE_U_INT64_T */
 
 /* Define to 1 if you have the <valgrind/valgrind.h> header file. */
-#define HAVE_VALGRIND_VALGRIND_H 1
+//#define HAVE_VALGRIND_VALGRIND_H 1
 
 /* Define to 1 if you have the `writev' function. */
 #define HAVE_WRITEV 1
@@ -699,7 +699,7 @@
 /* #undef const */
 
 /* Define to a type to use for `error_t' if it is not otherwise available. */
-#define error_t int
+//#define error_t int
 
 /* Define to `int' if <sys/types.h> does not define. */
 /* #undef pid_t */
