--- config.h	2013-11-23 14:11:37.000000000 -0700
+++ config.h	2013-11-23 14:11:37.000000000 -0700
@@ -38,7 +38,7 @@
 #define GCC_INSTALL_PREFIX ""

 /* Define to 1 if you have the `arc4random' function. */
-#define HAVE_ARC4RANDOM 1
+//#define HAVE_ARC4RANDOM 1

 /* Define to 1 if you have the `argz_append' function. */
 /* #undef HAVE_ARGZ_APPEND */
@@ -253,7 +253,7 @@
 #define HAVE_LONGJMP 1

 /* Define to 1 if you have the <mach/mach.h> header file. */
-#define HAVE_MACH_MACH_H 1
+//#define HAVE_MACH_MACH_H 1

 /* Define to 1 if you have the <mach-o/dyld.h> header file. */
 #define HAVE_MACH_O_DYLD_H 1
@@ -265,10 +265,10 @@
 /* #undef HAVE_MALLOC_H */

 /* Define to 1 if you have the <malloc/malloc.h> header file. */
-#define HAVE_MALLOC_MALLOC_H 1
+//#define HAVE_MALLOC_MALLOC_H 1

 /* Define to 1 if you have the `malloc_zone_statistics' function. */
-#define HAVE_MALLOC_ZONE_STATISTICS 1
+//#define HAVE_MALLOC_ZONE_STATISTICS 1

 /* Define to 1 if you have the <memory.h> header file. */
 #define HAVE_MEMORY_H 1
@@ -448,7 +448,7 @@
 #define HAVE_SYS_WAIT_H 1

 /* Define if the setupterm() function is supported this platform. */
-#define HAVE_TERMINFO 1
+//#define HAVE_TERMINFO 1

 /* Define to 1 if you have the <termios.h> header file. */
 #define HAVE_TERMIOS_H 1
@@ -717,7 +717,7 @@
 /* #undef const */

 /* Define to a type to use for `error_t' if it is not otherwise available. */
-#define error_t int
+//#define error_t int

 /* Define to `int' if <sys/types.h> does not define. */
 /* #undef pid_t */
