--- /Users/bfish/code/llvm/lib/Support/Unix/Process.inc	2016-10-14 12:38:41.000000000 -0700
+++ ./src/Support/Unix/Process.inc	2016-10-18 11:59:51.000000000 -0700
@@ -429,6 +429,7 @@
   return "\033[0m";
 }
 
+#if 0
 #if !HAVE_DECL_ARC4RANDOM
 static unsigned GetRandomNumberSeed() {
   // Attempt to get the initial seed from /dev/urandom, if possible.
@@ -463,3 +464,4 @@
   return ::rand();
 #endif
 }
+#endif //0
