--- ../../../llvm/lib/Support/Unix/Process.inc	2016-01-20 17:28:10.000000000 -0700
+++ ./src/Support/Unix/Process.inc	2016-02-07 11:44:53.000000000 -0700
@@ -427,6 +427,7 @@
   return "\033[0m";
 }
 
+#if 0
 #if !defined(HAVE_DECL_ARC4RANDOM) || !HAVE_DECL_ARC4RANDOM
 static unsigned GetRandomNumberSeed() {
   // Attempt to get the initial seed from /dev/urandom, if possible.
@@ -461,3 +462,4 @@
   return ::rand();
 #endif
 }
+#endif //0
