--- SourceMgr.cpp	2014-04-07 11:40:43.000000000 -0700
+++ SourceMgr.cpp	2014-04-09 12:57:26.000000000 -0700
@@ -297,8 +297,10 @@
     // FIXME: This assertion is intended to catch unintended use of multibyte
     // characters in fixits. If we decide to do this, we'll have to track
     // separate byte widths for the source and fixit lines.
+#if 0
     assert((size_t)llvm::sys::locale::columnWidth(I->getText()) ==
            I->getText().size());
+#endif
 
     // This relies on one byte per column in our fixit hints.
     unsigned LastColumnModified = HintCol + I->getText().size();
