--- ../../../llvm/lib/Support/SourceMgr.cpp	2016-01-20 17:28:10.000000000 -0700
+++ ./src/Support/SourceMgr.cpp	2016-01-22 12:35:08.000000000 -0700
@@ -15,7 +15,7 @@
 
 #include "llvm/Support/SourceMgr.h"
 #include "llvm/ADT/Twine.h"
-#include "llvm/Support/Locale.h"
+//#include "llvm/Support/Locale.h"
 #include "llvm/Support/MemoryBuffer.h"
 #include "llvm/Support/Path.h"
 #include "llvm/Support/raw_ostream.h"
@@ -285,8 +285,10 @@
     // FIXME: This assertion is intended to catch unintended use of multibyte
     // characters in fixits. If we decide to do this, we'll have to track
     // separate byte widths for the source and fixit lines.
+#if 0
     assert((size_t)llvm::sys::locale::columnWidth(I->getText()) ==
            I->getText().size());
+#endif //0
 
     // This relies on one byte per column in our fixit hints.
     unsigned LastColumnModified = HintCol + I->getText().size();
