--- SourceMgr.cpp	2013-10-11 10:40:22.000000000 -0700
+++ SourceMgr.cpp	2013-10-23 12:29:44.000000000 -0700
@@ -298,8 +298,10 @@
     // FIXME: This assertion is intended to catch unintended use of multibyte
     // characters in fixits. If we decide to do this, we'll have to track
     // separate byte widths for the source and fixit lines.
+#if 0
     assert((size_t)llvm::sys::locale::columnWidth(I->getText()) ==
            I->getText().size());
+#endif
 
     // This relies on one byte per column in our fixit hints.
     unsigned LastColumnModified = HintCol + I->getText().size();
