--- /Users/bfish/code/llvm/unittests/Support/YAMLParserTest.cpp	2016-10-14 12:38:40.000000000 -0700
+++ ./unittests/Support/YAMLParserTest.cpp	2016-10-18 11:39:28.000000000 -0700
@@ -16,6 +16,11 @@
 
 namespace llvm {
 
+// Allow llvm::StringRef to write to a std::ostream.
+inline std::ostream &operator<<(std::ostream &os, const StringRef &s) {
+  return os.write(s.data(), static_cast<std::streamsize>(s.size()));
+}
+
 static void SuppressDiagnosticsOutput(const SMDiagnostic &, void *) {
   // Prevent SourceMgr from writing errors to stderr
   // to reduce noise in unit test runs.
