--- YAMLTraits.cpp	2013-09-21 11:02:31.000000000 -0700
+++ YAMLTraits.cpp	2013-10-23 11:59:51.000000000 -0700
@@ -41,11 +41,19 @@
 //  Input
 //===----------------------------------------------------------------------===//
 
-Input::Input(StringRef InputContent, void *Ctxt)
-  : IO(Ctxt),
+Input::Input(StringRef InputContent, void *Ctxt, 
+             llvm::SourceMgr::DiagHandlerTy Handler, void *HandlerCtxt) //bfish
+  : IO(Ctxt), 
     Strm(new Stream(InputContent, SrcMgr)),
     CurrentNode(NULL) {
+  // Set handler _before_ parsing begins.
+  if (Handler) {
+    SrcMgr.setDiagHandler(Handler, HandlerCtxt);  //bfish
+  }
   DocIterator = Strm->begin();
+  if (Strm->failed()) {
+    EC = make_error_code(errc::invalid_argument); //bfish
+  }
 }
 
 Input::~Input() {
@@ -85,6 +93,11 @@
 void Input::beginMapping() {
   if (EC)
     return;
+  if (CurrentNode == nullptr) {             // bfish temp fix.
+    EC = make_error_code(errc::invalid_argument);
+    return;
+  }
+  assert(CurrentNode);        //bfish
   MapHNode *MN = dyn_cast<MapHNode>(CurrentNode);
   if (MN) {
     MN->ValidKeys.clear();
@@ -307,6 +320,11 @@
         memcpy(Buf, &StringStorage[0], Len);
         KeyStr = StringRef(Buf, Len);
       }
+      if (i->getValue() == nullptr) {       // bfish temp fix.          
+        EC = make_error_code(errc::invalid_argument);
+        break;
+      }
+      assert(i->getValue());   //bfish
       HNode *ValueHNode = this->createHNodes(i->getValue());
       if (EC)
         break;
