--- ../../../llvm_git/llvm/lib/Support/YAMLParser.cpp	2018-12-22 13:28:05.000000000 -0700
+++ ./src/Support/YAMLParser.cpp	2018-12-22 14:03:44.000000000 -0700
@@ -933,12 +933,13 @@
 }
 
 bool Scanner::consume(uint32_t Expected) {
-  if (Expected >= 0x80)
-    report_fatal_error("Not dealing with this yet");
+  assert(Expected < 0x80);
+  //if (Expected >= 0x80)
+  //  report_fatal_error("Not dealing with this yet");
   if (Current == End)
     return false;
-  if (uint8_t(*Current) >= 0x80)
-    report_fatal_error("Not dealing with this yet");
+  //if (uint8_t(*Current) >= 0x80)
+  //  report_fatal_error("Not dealing with this yet");
   if (uint8_t(*Current) == Expected) {
     ++Current;
     ++Column;
@@ -1393,6 +1394,12 @@
   Token T;
   T.Kind = Token::TK_Scalar;
   T.Range = StringRef(Start, Current - Start);
+
+  if (FlowLevel) {
+    // Make sure that plain scalars don't have trailing white space.
+    T.Range = T.Range.rtrim(" \n\r");
+  }
+
   TokenQueue.push_back(T);
 
   // Plain scalars can be simple keys.
@@ -2312,7 +2319,7 @@
   case Token::TK_Anchor:
     if (AnchorInfo.Kind == Token::TK_Anchor) {
       setError("Already encountered an anchor for this node!", T);
-      return nullptr;
+      return new (NodeAllocator) NullNode(stream.CurrentDoc);
     }
     AnchorInfo = getNext(); // Consume TK_Anchor.
     T = peekNext();
@@ -2320,7 +2327,7 @@
   case Token::TK_Tag:
     if (TagInfo.Kind == Token::TK_Tag) {
       setError("Already encountered a tag for this node!", T);
-      return nullptr;
+      return new (NodeAllocator) NullNode(stream.CurrentDoc);
     }
     TagInfo = getNext(); // Consume TK_Tag.
     T = peekNext();
@@ -2366,6 +2373,11 @@
                  , AnchorInfo.Range.substr(1)
                  , TagInfo.Range
                  , MappingNode::MT_Flow);
+  case Token::TK_FlowSequenceEnd:
+  case Token::TK_FlowMappingEnd:
+  case Token::TK_FlowEntry:
+    setError("Unexpected token", T);
+    return new (NodeAllocator) NullNode(stream.CurrentDoc);
   case Token::TK_Scalar:
     getNext();
     return new (NodeAllocator)
@@ -2396,7 +2408,8 @@
     //       !!null null.
     return new (NodeAllocator) NullNode(stream.CurrentDoc);
   case Token::TK_Error:
-    return nullptr;
+    setError("Unexpected token", T);
+    return new (NodeAllocator) NullNode(stream.CurrentDoc);
   }
   llvm_unreachable("Control flow shouldn't reach here.");
   return nullptr;
