--- ../../../llvm_git/llvm/unittests/Support/YAMLIOTest.cpp	2018-12-22 13:28:13.000000000 -0700
+++ ./unittests/Support/YAMLIOTest.cpp	2018-12-22 14:03:59.000000000 -0700
@@ -14,7 +14,7 @@
 #include "llvm/Support/Endian.h"
 #include "llvm/Support/Format.h"
 #include "llvm/Support/YAMLTraits.h"
-#include "gmock/gmock.h"
+//#include "gmock/gmock.h"
 #include "gtest/gtest.h"
 
 using llvm::yaml::Hex16;
@@ -28,7 +28,7 @@
 using llvm::yaml::MappingTraits;
 using llvm::yaml::Output;
 using llvm::yaml::ScalarTraits;
-using ::testing::StartsWith;
+//using ::testing::StartsWith;
 
 
 
@@ -317,7 +317,7 @@
   }
   auto ExpectedPrefix = "---\n"
                         "str1:            aba\tcaba\n";
-  EXPECT_THAT(Serialized, StartsWith(ExpectedPrefix));
+  EXPECT_TRUE(llvm::StringRef{Serialized}.startswith(ExpectedPrefix));
 }
 
 //===----------------------------------------------------------------------===//
